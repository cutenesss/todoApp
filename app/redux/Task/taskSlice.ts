import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { ItemTaskProps } from '../../models/TaskModels'

export interface ProductState {
  listTask: Array<ItemTaskProps>
}

const initialState: ProductState = {
  listTask: [],
}

interface editTaskBody {
  idOldItem: number
  newItem: ItemTaskProps
}

const TaskSlice = createSlice({
  name: 'task',
  initialState,
  reducers: {
    addTask(state, action: PayloadAction<ItemTaskProps>) {
      state.listTask = state.listTask.concat(action.payload)
    },
    editTask(state, action: PayloadAction<editTaskBody>) {
      const currentItemIndex = state.listTask.findIndex(
        task => task.id === action.payload.idOldItem
      )
      const newListTask = [...state.listTask]
      newListTask[currentItemIndex].description =
        action.payload.newItem.description
      newListTask[currentItemIndex].name = action.payload.newItem.name
      state.listTask = newListTask
    },
  },
})

export const { addTask, editTask } = TaskSlice.actions
export default TaskSlice.reducer
