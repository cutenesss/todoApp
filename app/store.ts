import { configureStore, PreloadedState } from '@reduxjs/toolkit'
import RootReducer from './rootReducer'
import Reactotron from './config/ReactotronConfig'
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { useDispatch, TypedUseSelectorHook, useSelector } from 'react-redux'
import DebugConfig from './config/DebugConfig'

// const enhancers = DebugConfig.reactotron ? [Reactotron.createEnhancer!()] : []

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: RootReducer,
    middleware: getDefaultMiddleware => getDefaultMiddleware(),
    preloadedState
  // enhancers,
  })
}

export type RootState = ReturnType<typeof RootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']
export const useAppDispatch: () => AppDispatch = useDispatch
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
