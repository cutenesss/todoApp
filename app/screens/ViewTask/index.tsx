import React, { useCallback, useRef } from 'react'
import { Text, StyleSheet, TouchableOpacity, View } from 'react-native'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import { addTask, deleteTask, completeTask } from '../../redux'
import { useAppDispatch } from '../../store'
import { MInput } from '../../components'
import { RootStackParamList } from '../../models/NavigationType'
import { PRIORITY_LEVEL } from '../../constant';
import { RadioButtonGroup } from '../../components'
import { getPriorityLevel, onAlert } from '../../utils'

export type ViewTaskScreenProps = {
    navigation: NativeStackScreenProps<RootStackParamList, 'HOME'>['navigation'],
    route?: NativeStackScreenProps<RootStackParamList, 'VIEW_TASK'>['route']
}

const ViewTaskScreen = ({ navigation, route }: ViewTaskScreenProps) => {
    const itemTask = route?.params?.item
    const dispatch = useAppDispatch()
    const nameRef = useRef<any>()
    const descriptionRef = useRef<any>()
    const priorityRef = useRef<PRIORITY_LEVEL | ''>('')
    const listPriorityLevel = [
        {
            id: 'priority_1',
            label: getPriorityLevel(PRIORITY_LEVEL.HIGH),
            value: PRIORITY_LEVEL.HIGH,
            selected: false
        },
        {
            id: 'priority_2',
            label: getPriorityLevel(PRIORITY_LEVEL.NORMAL),
            value: PRIORITY_LEVEL.NORMAL,
            selected: false

        },
        {
            id: 'priority_3',
            label: getPriorityLevel(PRIORITY_LEVEL.LOW),
            value: PRIORITY_LEVEL.LOW,
            selected: false
        }
    ]
    
    const returnAfterSuccess = () => navigation.goBack()

    // these function has been tested inside the component
    const resetText = useCallback(() => {
        /* istanbul ignore next */
        nameRef.current.resetText()
        /* istanbul ignore next */
        descriptionRef.current.resetText()
        priorityRef.current = ''
    }, [])

    const handleItem = () => {
        if (itemTask) {
            dispatch(
                deleteTask({ idItem: itemTask?.id })
            )
            onAlert({
                title: "Delete success",
                okFunction: returnAfterSuccess
            })
        } else {
            if (nameRef.current.getText() === '') {
                onAlert({ title: "Enter task's name" })
            } else {
                dispatch(
                    addTask({
                        id: Date.now(),
                        name: nameRef.current.getText(),
                        description: descriptionRef.current.getText(),
                        priorityLevel: priorityRef.current === '' ? PRIORITY_LEVEL.NORMAL : priorityRef.current,
                        isCompleted: false
                    })
                )
                /* istanbul ignore next */
                resetText()
                onAlert({
                    title: 'Add success',
                    okFunction: returnAfterSuccess
                })
            }
        }
    }

    const handleCompleteItem = () => {
        if (itemTask?.id) {
            dispatch(
                completeTask({ idItem: itemTask?.id })
            )
            onAlert({
                title: "Completed task",
                okFunction: returnAfterSuccess
            })
        }
    }

    const onPressRadioButton = (value: PRIORITY_LEVEL) => {
        priorityRef.current = value
    }

    return (
        <View style={styles.container}>
            <MInput
                testId="Name"
                placeholder="Name"
                containerStyle={styles.input}
                ref={nameRef}
                defaultText={itemTask?.name ?? ''}
                editable={itemTask?.name ? false : true}
            />
            <MInput
                testId="Description"
                placeholder="Description"
                containerStyle={styles.input}
                ref={descriptionRef}
                defaultText={itemTask?.description ?? ''}
                editable={itemTask?.description ? false : true}
            />
            <RadioButtonGroup
                title='Priority'
                initList={listPriorityLevel}
                onPress={onPressRadioButton}
                initValue={itemTask?.priorityLevel}
                disabled={itemTask?.id ? true : false}
            />
            {
                itemTask && (
                    <TouchableOpacity onPress={handleCompleteItem} style={styles.button}>
                        <Text style={styles.buttonText}>Complete</Text>
                    </TouchableOpacity>
                )
            }
            <TouchableOpacity onPress={handleItem} style={styles.button}>
                <Text style={styles.buttonText}>{itemTask ? 'Delete' : 'Add'}</Text>
            </TouchableOpacity>
        </View>
    )
}

export default ViewTaskScreen

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
        paddingHorizontal: 14,
    },
    button: {
        backgroundColor: 'blue',
        borderRadius: 10,
        padding: 10,
        width: '100%',
        marginTop: 20,
        alignItems: 'center',
        justifyContent: 'center',
    },
    buttonText: {
        color: '#fff',
        fontSize: 16,
    },
    input: { marginTop: 15 },
})
