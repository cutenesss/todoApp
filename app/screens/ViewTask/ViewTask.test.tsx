import React from 'react';
import { Alert } from 'react-native';
import renderer from 'react-test-renderer';
import { render, fireEvent, } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { setupStore } from '../../store';
import { testItemSlice } from '../../redux/Task/taskSlice.test';
import ViewTask from '.';

const ID_INPUT_NAME = 'MInput_input_Name'
const ID_INPUT_DESCRIPTION = 'MInput_input_Description'

const testRoute: any = {
    params: {
        item: testItemSlice
    }
}

const navigateMock = jest.fn();
const navigation: any = {
    navigate: navigateMock
};

describe("<ViewTask />", () => {
    test('renders with no item correctly', () => {
        const tree = renderer
            .create(
                <Provider store={setupStore()}>
                    <ViewTask navigation={navigation} />
                </Provider>
            )
            .toJSON()
        expect(tree).toMatchSnapshot();
    });

    test('renders with item correctly', () => {
        const tree = render
            (
                <Provider store={setupStore()}>
                    <ViewTask navigation={navigation} route={testRoute} />
                </Provider>
            )
        expect(tree.getByTestId(ID_INPUT_NAME).props.value).toBe('name');
        expect(tree.getByTestId(ID_INPUT_DESCRIPTION).props.value).toBe('description');
    });

    test('add item correctly', () => {
        const { getByTestId, getByText } = render
            (
                <Provider store={setupStore()}>
                    <ViewTask navigation={navigation} />
                </Provider>
            )
        const inputName = getByTestId(ID_INPUT_NAME);
        expect(getByTestId(ID_INPUT_NAME).props.value).toBe('')
        fireEvent.changeText(inputName, 'name');
        expect(getByTestId(ID_INPUT_NAME).props.value).toBe('name')

        const inputDescription = getByTestId(ID_INPUT_DESCRIPTION);
        expect(getByTestId(ID_INPUT_DESCRIPTION).props.value).toBe('')
        fireEvent.changeText(inputDescription, 'description');
        expect(getByTestId(ID_INPUT_DESCRIPTION).props.value).toBe('description')
        jest.spyOn(Alert, 'alert');
        fireEvent.press(getByText('Add'));
        expect(Alert.alert).toHaveBeenCalledWith('Add success')
    });

    test('add item without name', () => {
        const { getByTestId, getByText } = render
            (
                <Provider store={setupStore()}>
                    <ViewTask navigation={navigation} />
                </Provider>
            )
        const inputDescription = getByTestId(ID_INPUT_DESCRIPTION);
        expect(getByTestId(ID_INPUT_DESCRIPTION).props.value).toBe('')
        fireEvent.changeText(inputDescription, 'description');
        expect(getByTestId(ID_INPUT_DESCRIPTION).props.value).toBe('description')
        jest.spyOn(Alert, 'alert');
        fireEvent.press(getByText('Add'));
        expect(Alert.alert).toHaveBeenCalledWith("Enter task's name")
    });

    test('delete item correctly', () => {
        const initialState = {
            task: {
                listTask: [testItemSlice],
                completedTask: 0
            }
        };
        const store = setupStore(initialState)
        const { getByTestId, getByText } = render
            (
                <Provider store={store}>
                    <ViewTask navigation={navigation} route={testRoute} />
                </Provider>
            )
        expect(getByTestId(ID_INPUT_NAME).props.value).toBe('name');
        expect(getByTestId(ID_INPUT_DESCRIPTION).props.value).toBe('description');
    });
});
