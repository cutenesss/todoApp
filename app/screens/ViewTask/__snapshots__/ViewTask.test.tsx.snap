// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewTask /> renders with no item correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "paddingHorizontal": 14,
    }
  }
>
  <View
    style={
      Object {
        "marginTop": 15,
      }
    }
    testID="MInput_view_Name"
  >
    <TextInput
      editable={true}
      onChangeText={[Function]}
      placeholder="Name"
      secureTextEntry={false}
      style={
        Object {
          "borderColor": "#D9D9D9",
          "borderRadius": 5,
          "borderWidth": 1,
          "fontSize": 16,
          "minHeight": 40,
          "paddingHorizontal": 10,
        }
      }
      testID="MInput_input_Name"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "marginTop": 15,
      }
    }
    testID="MInput_view_Description"
  >
    <TextInput
      editable={true}
      onChangeText={[Function]}
      placeholder="Description"
      secureTextEntry={false}
      style={
        Object {
          "borderColor": "#D9D9D9",
          "borderRadius": 5,
          "borderWidth": 1,
          "fontSize": 16,
          "minHeight": 40,
          "paddingHorizontal": 10,
        }
      }
      testID="MInput_input_Description"
      value=""
    />
  </View>
  <View>
    <Text
      style={
        Object {
          "marginBottom": 4,
          "marginTop": 16,
        }
      }
    >
      Priority
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
          },
          Object {
            "flexDirection": "column",
          },
          Object {
            "alignItems": "flex-start",
          },
        ]
      }
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "blue",
        "borderRadius": 10,
        "justifyContent": "center",
        "marginTop": 20,
        "opacity": 1,
        "padding": 10,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 16,
        }
      }
    >
      Add
    </Text>
  </View>
</View>
`;
